tokio-trace/src/lib.rs

```diff
@@ -0,0 +1,780 @@
+            // Depending on how many fields are generated, this may or may
+            // not actually be used, but it doesn't make sense to repeat it.
+            #[allow(unused_variables, unused_mut)] {
+                if !span.is_disabled() {
+                    let mut keys = callsite.metadata().fields().into_iter();
+                    $(
+                        let key = keys.next()
+                            .expect(concat!("metadata should define a key for '", stringify!($k), "'"));
+                        span!(@ record: span, $k, &key, $($val)*);
+                    )*
+                };
+            }
+            span
+        }
+    };
+    (@ record: $span:expr, $k:expr, $i:expr, $val:expr) => (
+        $span.record($i, &$val)
+    );
+    (@ record: $span:expr, $k:expr, $i:expr,) => (
+        // skip
+    );
+}
+
+/// Constructs a new `Event`.
+///
+/// # Examples
+///
+/// ```rust
+/// # #[macro_use]
+/// # extern crate tokio_trace;
+/// use tokio_trace::{Level, field};
+///
+/// # fn main() {
+/// let data = (42, "fourty-two");
+/// let private_data = "private";
+/// let error = "a bad error";
+///
+/// event!(Level::ERROR, { error = field::display(error) }, "Received error");
+/// event!(target: "app_events", Level::WARN, {
+///         private_data = private_data,
+///         data = field::debug(data)
+///     },
+///     "App warning: {}", error
+/// );
+/// event!(Level::INFO, the_answer = data.0);
+/// # }
+/// ```
+#[macro_export]
+macro_rules! event {
+    (target: $target:expr, $lvl:expr, $( $k:ident $( = $val:expr )* ),+ ) => (
+        event!(target: $target, $lvl, { $( $k $( = $val)* ),+ })
+    );
+    (target: $target:expr, $lvl:expr, { $( $k:ident $( = $val:expr )* ),* }, $($arg:tt)+ ) => ({
+        {
+            #[allow(unused_imports)]
+            use $crate::{callsite, Id, Subscriber, Event, field::{Value, AsField}};
+            use $crate::callsite::Callsite;
+            let callsite = callsite! { event:
+                $lvl,
+                target:
+                $target, $( $k ),*
+            };
+            let mut event = Event::new(callsite.interest(), callsite.metadata());
+            // Depending on how many fields are generated, this may or may
+            // not actually be used, but it doesn't make sense to repeat it.
+            #[allow(unused_variables, unused_mut)] {
+                if !event.is_disabled() {
+                    let mut keys = callsite.metadata().fields().into_iter();
+                    let msg_key = keys.next()
+                        .expect("event metadata should define a key for the message");
+                    event.message(&msg_key, format_args!( $($arg)+ ));
+                    $(
+                        let key = keys.next()
+                            .expect(concat!("metadata should define a key for '", stringify!($k), "'"));
+                        event!(@ record: event, $k, &key, $($val)*);
```

Comment:
I think this could be tricky for frameworks like `slog` (and maybe also `log`) that expect the structured data to be wholly and readily accessible rather than streamed one value at a time. While values are either `Debug` or one of a few primitive types this probably isn't a problem to work around, but might be something we should think about as the value API evolves?

In my opinion, having to box values for integrating with other frameworks is ok.

Reply:
Thanks for your feedback --- compatibility with other frameworks here is definitely an important consideration.

> I think this could be tricky for frameworks like slog (and maybe also log) that expect the structured data to be wholly and readily accessible rather than streamed one value at a time.

I think an adapter for a framework like `slog` could solve this problem by waiting for all the field values to be added prior to exposing an event to `slog` --- this is part of the motivation behind why `tokio-trace` stipulates that all the fields on an event must be _declared_ when the event is created, even if their values are not provided at that time. This way,  we know when the event's fields are "complete". The adapter would have to serialize/format all the field values individually as they come in before exposing them to the other framework. I don't _think_ that's an issue but I could be incorrect.

> In my opinion, having to box values for integrating with other frameworks is ok.

I'm also fine with having integration code/adapters allocate. However, @carllerche has stressed that ensuring the instrumentation code should never allocate per-field or per-event is a major design goal for `tokio-trace`. We should ensure that the choice of whether or not to allocate here is left to the _consumer_ of the trace data.

Reply:
> The adapter would have to serialize/format all the field values individually as they come in before exposing them to the other framework.

That's kind of what I was thinking too. Capture values into a `serde_json::Value`-like thing. If we get to the point of designing the record API for maps and sequences then getting a chance to box an anonymous value might make that serialization a bit cheaper.

> We should ensure that the choice of whether or not to allocate here is left to the consumer of the trace data.

That seems fair. It's reasonable for `tokio-trace` to optimise its design around being the sole provider of events in an application, and work to keep that path as allocation free as possible. For integration points where paradigms meet we'll probably have to accept some cost.

So all that's to say I like the way this works.

Reply:
> So all that's to say I like the way this works.

That's good to hear, thanks! :D